layer/layer_store.go

```diff
@@ -409,7 +410,13 @@ func (ls *layerStore) Map() map[ChainID]Layer {
 }
 
 func (ls *layerStore) deleteLayer(layer *roLayer, metadata *Metadata) error {
-	err := ls.driver.Remove(layer.cacheID)
+	// Rename layer digest folder first so we detect orphan layer(s)
+	// if ls.driver.Remove fails
+	err := ls.store.PrepareRemove(layer.chainID)
+	if err != nil {
+		return err
+	}
+	err = ls.driver.Remove(layer.cacheID)
 	if err != nil {
 		return err
```

Comment:
Should the layer be restored in this case?

Reply:
No graph drivers removes layer from disk (or try at least) and when first file is removed from disk layer is not valid anymore and it should not be used for anything.

Reply:
Seems like a mismatch in how atomic operations are. A partial delete should be restarted? What happens when the operation is retried?

Reply:
Windows graphdriver already tries to do so https://github.com/moby/moby/blob/ea60a87fcfc8b4592a5d2dec60f021c44ebe6a05/daemon/graphdriver/windows/windows.go#L334-L347 but it cannot be trusted. Only way get those layers fully released is reboot whole server and let ImageService.Cleanup() to remove those.

Image is removed from imageDB first so if user will try remove it again he will get error that image does not exist (even without this fix).